docs/Langref.md

```diff
@@ -0,0 +1,375 @@
+# Iridium Language Reference
+## Terminologies
+### Value
+A value is a piece of data that can be represented by a finite number of bits.
+### Type
+A type is a possible set of values.
+### Operator
+An operator is a function that takes one or more values and returns a value or a value tuple. It can have a sub-region optionally.
+### Value
+A value is the result of an operator or a value passed to an operator/a block.
+### Attributes
+Attributes describes the metadata of a value (range/distribution/TBAA). It can be attached to an operand and propagated along the use-def chain. Attributes that can be attached are determined by the type of the operand. It is widely used in the flow-sensitive data analysis.
+### Metadata
+Metadata provides additional information about.
+### Block
+A block is a DAG of operators. It takes an entry state and some arguments. A block must end with an operator which implements the `Terminator` trait.
+### Region
+A region is a set of blocks. It also maintains the control flow graph (CFG) and the dominator tree (DT) of these blocks.
+### State
+A state models all side effects of an operator. It can be passed to an operator as an operand and returned by an operator. It determines the execution order of operators (e.g., a load must be executed after a store). Thus we can easily track them in our data analysis framework.
+## Undefined Behavior
+### Poison Value
+A poison value represents a violation of an assumption. Each operation either propagates `poison` or raise UB.
+### Valid Transformation
+A transformation is valid when, for each possible input state, the target function displays a subset of the behaviors of the source function. It allows a transformation to remove non-determinism, but not to add it.
+
+## Type System
+### Scalars
+#### Integers
+##### Boolean
+A boolean value can be either `true` or `false`. It is often used by conditional branch operators.
+Available attributes: `probabilty`, `no_poison`
+Avaliable operators: `and`, `or`, `xor`, `logical_and`, `logical_or`
+##### Signed Integer
+A signed integer represents a integer value in range `[-2^(n-1), 2^(n-1)-1]`, where `n` is the bit width of the integer. The result of a signed integer is a poison value if signed overflow occurs.
```

Comment:
```suggestion
A signed integer represents a integer value in range `[-2^(n-1), 2^(n-1)-1]`, where `n` is the bit width of the integer. 
```

Poison value should only be produced by *operations*, I think we should state "poison" in op definition.